### PR TITLE
Implement `DataValidator.apply()`

### PR DESCRIPTION
**Pull request by danielhuppmann**
This is the implementation for the `apply` method of the `DataValidator`.

The current implementation writes each validation-item to the log separately, together with the failing data rows.

```
2024-08-19 10:10:33 ERROR    Failed data validation (file validate_data/validate_data_fails.yaml):
  Criteria: variable: ['Primary Energy'], upper_bound: 5.0
         model scenario region        variable   unit  year  value
    0  model_a   scen_a  World  Primary Energy  EJ/yr  2010    6.0
    1  model_a   scen_b  World  Primary Energy  EJ/yr  2010    7.0

  Criteria: variable: ['Primary Energy|Coal'], lower_bound: 2.0
         model scenario region             variable   unit  year  value
    0  model_a   scen_a  World  Primary Energy|Coal  EJ/yr  2005    0.5

  Criteria: variable: ['Primary Energy'], year: [2005], upper_bound: 1.9, lower_bound: 1.1
         model scenario region        variable   unit  year  value
    0  model_a   scen_a  World  Primary Energy  EJ/yr  2005    1.0
    1  model_a   scen_b  World  Primary Energy  EJ/yr  2005    2.0
```

I initially had all failing validations (per yaml file) concatenated to one DataFrame, but I then thought that it would not be helpful if users can't see which criteria they specifically do not pass.

If you agree with this approach in principle, I can make the error-message a bit nicer (indentation, more concise criteria representation).

### CODE REVIEW

**Review by phackstock — APPROVED** [review]
Looks good to me, some smaller comments in line and one file to delete (I think).
After that, good to merge.

File: tests/data/validation/Untitled.ipynb
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply tests/data/validation/Untitled.ipynb:1]: Sorry - removed the file

File: nomenclature/processor/data_validator.py [hunk 1]
```diff
@@ -29,14 +35,39 @@ def from_file(cls, file: Union[Path, str]) -> "DataValidator":
...
+                )
+                raise ValueError(
+                    "Data validation failed. Please check the log for details."
+                )
+        return df
 
     def validate_with_definition(self, dsd: DataStructureDefinition) -> None:
         errors = ErrorCollector(description=f"in file '{self.file}'")
-        for data in self.criteria_items:
+        for item in self.criteria_items:
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply nomenclature/processor/data_validator.py]: Thanks, implemented

File: nomenclature/processor/data_validator.py [hunk 2]
```diff
@@ -29,14 +35,39 @@ def from_file(cls, file: Union[Path, str]) -> "DataValidator":
             content = yaml.safe_load(f)
         return cls(file=file, criteria_items=content)
 
-    def apply(self):
-        pass
+    def apply(self, df: IamDataFrame) -> IamDataFrame:
+        error_list = []
+
+        with adjust_log_level():
+            for item in self.criteria_items:
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply nomenclature/processor/data_validator.py:42]: This seems less readable than the current implementation, so suggest to get it unless we make a utility-function that moves the string-concat somewhere else - maybe together with a refactor of the `RequiredDataValidator`?

**Pull request merged**